File: sus_scriptsTruck1.0.14/Assets/Scripts/Lua/Core/Res/ResPreLoader.lua
local class = Core.oop.class;
local ResPreLoader = class("ResPreLoader")
local M = ResPreLoader

function M:ctor(resTag)
    self.preLoadTasks = {}
    self.isPreLoad = false
    self.resTag = resTag
end

function M:Begin()
    self.isPreLoad = true
end

function M:End()
    self.isPreLoad = false
end

function M:Progress()
    if not self.isPreLoad then
        return 1
    end
    local p = 0
    for k,v in pairs(self.preLoadTasks) do
        p = p + v:Progress()
    end
    p = p/#self.preLoadTasks
    return p
end

function M:IsDone()
    if not self.isPreLoad then
        return true
    end
    local isDone = true
    for k,v in pairs(self.preLoadTasks) do
        if not v:IsDone() then
            isDone = false
            break
        end
        v:DoneCallback()
    end
    return isDone
end

function M:Push(resType, name, callback)
    local task = Logic.res.LoadAsync(self.resTag, resType, name, callback)
    if not task then
        Logic.log.Error("预加载失败:type=" .. resType .. "," .. name)
    end
    task:Retain(self.resTag)
    table.insert(self.preLoadTasks, task)
    return task
end

return M

File: sus_scriptsTruck1.0.14/Assets/Scripts/CSharp/UI/BookReading/ResPreLoader.cs
#if NOT_USE_LUA
using AB;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace GameLogic
{
    public class ResPreLoader
    {
        List<CAsset> list = new List<CAsset>();
        public Dictionary<string,CAsset> lstTask = new Dictionary<string, CAsset>();
        public bool isPreLoad = false;
        [NonSerialized]
        public int hadLoaded = 0;

        public void BeginPreLoad()
        {
            isPreLoad = true;
        }

        public void EndPreLoad()
        {
            isPreLoad = false;
        }


        public float GetPreLoadProgress()
        {
            if (!isPreLoad)
            {
                return 0;
            }

            float p = 0;
            for(int i= 0; i < list.Count;++i)
            {
                var task = list[i];
                p += task.Progress();
            }
            p /= list.Count;
            return p;
        }

        public bool IsPreLoadDone()
        {
            if (!isPreLoad)
            {
                return false;
            }

            hadLoaded = 0;
            bool isDone = true;
            for (int i = 0; i < list.Count; ++i)
            {
                var task = list[i];
                if (!task.IsDone())
                {
                    isDone = false;
                    break;
                }
                task.DoneCallback();
                ++hadLoaded;
            }
            return isDone;
        }

        public CAsset PreLoad(enResType resType, string strAssetName, Action<CAsset> finishFunc = null)
        {
            CAsset task;
            //if (this.lstTask.TryGetValue(strAssetName, out task))
            //{
            //    if (task.IsDone())
            //    {
            //        if(task.m_resType == enResType.eObject && task.m_resType == resType)//Object默认加载出来的是Texture2D
            //        {
            //            if (finishFunc != null)
            //            {
            //                finishFunc((CAsset)task);
            //            }
            //            return task;
            //        }

            //    }
            //    else
            //    {
            //        //this.lstTask.Remove(strAssetName);//有可能是空的
            //    }
            //}

            task = ABSystem.Instance.LoadAsync(AbTag.Null, resType, strAssetName, finishFunc);
            if (task == null)
            {
                LOG.Error("预加载失败:type=" + resType + "," + strAssetName);
                return null;
            }
            task.Retain(this);
            this.lstTask[strAssetName] = task;
            list.Add(task);
            return task;
        }

        public void Clean()
        {
            for (int i = 0; i < list.Count; ++i)
            {
                var task = list[i];
                task.Release(this);
            }
            lstTask.Clear();
            list.Clear();
        }
    }
}
#endif
